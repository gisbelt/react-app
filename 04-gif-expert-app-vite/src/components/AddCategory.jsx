import { useState } from "react"

// export const AddCategory = ( {setCategories} ) => {
export const AddCategory = ( {onNewValue} ) => {

    const [inputValue, setInputValue] = useState('')
    // function to change the input value
    const onInputChange = (event) => {
        setInputValue( event.target.value )
    }

    const onSubmit = ( event ) => {
        // prevent the browser from refreshing when pressing enter
        event.preventDefault();
        // validate that the input is not empty.
        if ( inputValue.trim().length <= 1 ) return;
               
        // Better way, send new value through props
        onNewValue( inputValue.trim() )

        // clean the input
        setInputValue('')
    }

    return (
        
        <form onSubmit={ (event) => onSubmit(event) }>
            <input 
                type="text"
                placeholder="search gifs"
                value={ inputValue }
                onChange={ onInputChange }
            />
        </form>
    )
}

