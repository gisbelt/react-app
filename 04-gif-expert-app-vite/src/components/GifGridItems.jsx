import React from 'react'
import Proptypes from "prop-types";

export const GifGridItems = ({title, url}) => {
    console.log(title, url)
    return (
        <div className='card'>
            <img src={ url } alt={ title } />
            <p>{ title }</p>
        </div> 
    )
}

GifGridItems.propTypes = {
    title: Proptypes.string.isRequired,
    url: Proptypes.string.isRequired
}