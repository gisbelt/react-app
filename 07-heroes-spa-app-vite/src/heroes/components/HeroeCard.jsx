import { Link } from "react-router-dom"

export const HeroeCard = ({ id, superhero, publisher, alter_ego, first_appearance, characters }) => {
    
    // we receive all images 
    const heroImageUrl = `/assets/heroes/${id}.jpg`

    return (
        <div className="col">
            <div className="card text-start h-100">
                <div className="row g-0">
                    <div className="col-4">
                        <img 
                            className="card-img-top h-100" 
                            src={ heroImageUrl }    
                            alt={ superhero } 
                        />
                    </div>
                
                    <div className="col-8">
                        <div className="card-body">
                            <h4 className="card-title">{ superhero }</h4>
                            <p className="card-text">{ alter_ego }</p>
                            { (alter_ego !== characters ) && (<p className="card-text">{ characters }</p>) }
                            <p className="card-text">
                                <small className="text-muted">{ first_appearance }</small>
                            </p>

                            <Link to={`/hero/${id}`}>
                                View more...
                            </Link>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
}
