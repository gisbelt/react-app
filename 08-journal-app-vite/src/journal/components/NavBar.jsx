import { MenuOutlined } from "@mui/icons-material"
import { AppBar, Avatar, Button, Grid, IconButton, Menu, MenuItem, Toolbar, Tooltip, Typography } from "@mui/material"
import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { startLogout } from "../../store/auth/thunks";

const settings = [
	{ title: "Profile", path: "/profile" },
	{ title: "Account", path: "/account" },
	{ title: "Dashboard", path: "/dashboard" },
];

export const NavBar = ({ drawerWidth }) => {

	const dispatch = useDispatch()
	const { photoURL } = useSelector(state => state.auth )

  	const [anchorElUser, setAnchorElUser] = useState(null);

	const handleOpenUserMenu = (event) => {
		setAnchorElUser(event.currentTarget);
	};

	const handleCloseUserMenu = () => {
		setAnchorElUser(null);
	};

	const handleLogout = () => {
		dispatch( startLogout() );
	}

	return (
		<AppBar
			position="fixed"
			sx={{ 
				width: {sm: `calc(100% - ${drawerWidth}px)`},
				ml: { sm: `${drawerWidth}px`}
			}}
		>
			<Toolbar>
				<IconButton
					size="large"
					edge="start"
					color="inherit"
					aria-label="menu"
					sx={{ mr: 2, display: { sm: 'none' }}}
				>
					<MenuOutlined />
				</IconButton>
			

				<Grid container direction='row' justifyContent='space-between' alignItems='center'>

					<Typography variant="h6" noWrap component="div" sx={{ flexGrow: 1 }}>
						Journal App
					</Typography>

					<Tooltip title="Open settings">
						<IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
							<Avatar alt="Avatar" src={`${photoURL}`} />
						</IconButton>
					</Tooltip>
					<Menu
						sx={{ mt: '45px' }}
						id="menu-appbar"
						anchorEl={anchorElUser}
						anchorOrigin={{
						  vertical: 'top',
						  horizontal: 'right',
						}}
						keepMounted
						transformOrigin={{
						  vertical: 'top',
						  horizontal: 'right',
						}}
						open={Boolean(anchorElUser)}
						onClose={handleCloseUserMenu}
					>
						{settings.map((setting, index) => (
							<MenuItem key={index} onClick={handleCloseUserMenu}>
								<Link to={ setting.path } >{setting.title}</Link>
							</MenuItem>
						))}
						<MenuItem onClick={handleCloseUserMenu}>
							<Link type="button" onClick={ handleLogout }>Logout</Link>
						</MenuItem>
					</Menu>
				</Grid>
			</Toolbar>
		</AppBar>
	)
}
