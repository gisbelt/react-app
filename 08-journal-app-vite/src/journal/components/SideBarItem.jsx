import { TurnedInNot } from '@mui/icons-material'
import { Grid, ListItem, ListItemButton, ListItemIcon, ListItemText } from '@mui/material'
import React, { useMemo } from 'react'

export const SideBarItem = ({ title, body, id }) => {

	const newTitle = useMemo( () => {
		return title.length > 10
			? title.substring(0,10) + '...'
			: title;
	}, [ title ])

	return (
		<ListItem disablePadding>
			<ListItemButton>
				<ListItemIcon>
					<TurnedInNot />
				</ListItemIcon>
				<Grid container>
					<ListItemText primary={ newTitle }/>
					<ListItemText secondary={ body }/>
				</Grid>
			</ListItemButton>
		</ListItem>
	)
}
