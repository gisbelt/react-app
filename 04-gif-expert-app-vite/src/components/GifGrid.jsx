import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItems } from "./GifGridItems";
 
export const GifGrid = ( { category } ) => {

  //Custom Hook
  const { images, isLoading } = useFetchGifs( category );
  
  return (
    <>
        <h3>{ category }</h3>
        {
          isLoading && (<h2>Loading...</h2>)
        }
        <div className="card-grid">
          {
            images.map( (images) => (
              <GifGridItems 
                key={images.id}
                {...images} //spread operator to get all the props from images
              />
            ))
          }
        </div>
    </>
  )
}

