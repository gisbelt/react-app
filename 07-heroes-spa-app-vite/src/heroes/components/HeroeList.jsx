import React from 'react'
import { getHeroesByPublisher } from '../helpers/getHeroesByPublisher'
import { HeroeCard } from './HeroeCard'

export const HeroeList = ({ publisher }) => {

    const heroes = getHeroesByPublisher(publisher)
    
    return (
        <>
            <div className='row row-cols-1 row-cols-md-3 g-3'>
                {
                    heroes.map((item, index) => (
                        <HeroeCard 
                            key={ item.id }
                            {... item} //spread operator to receive props
                        />
                    ))
                }
            </div>
        </>
    )
}
