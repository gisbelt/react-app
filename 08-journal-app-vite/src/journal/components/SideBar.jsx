import { Box, Divider, Drawer, List, Toolbar, Typography } from "@mui/material"
import { useSelector } from "react-redux";
import { SideBarItem } from "./SideBarItem";

export const SideBar = ({ drawerWidth }) => {

	const { displayName } = useSelector(state => state.auth )
	const { notes } = useSelector(state => state.journal )
	
	return (
		<Box
			component='nav'
			sx={{ width: { sm: `${drawerWidth}px` }, flexShrink: { sm: 0 } }}
		>
			<Drawer
				variant='permanent' // temporary
				open
				// onClose 
				// ModalProps={{
				// 	keepMounted: true,
				// }}
				sx={{ 
					xs: 'block', 
					'& .MuiDrawer-paper': { boxSizing: 'border-box', width: drawerWidth } 
				}}
			>
				<Toolbar>
					<Typography variant="h6" noWrap component='div' >{ displayName }</Typography>
				</Toolbar>
				<Divider />

				<List>
					{
						notes.map((note, index) => (
							<SideBarItem key={ index } { ...note } />
						))
					}
				</List>
			</Drawer>
		</Box>
	)
}
