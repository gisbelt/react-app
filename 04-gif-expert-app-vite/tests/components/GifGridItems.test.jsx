import { render, screen } from "@testing-library/react"
import { GifGridItems } from "../../src/components/GifGridItems";

describe('Testing component <GifGridItems />', () => { 
    const title = 'DragonBall'
    const url = 'https://api.giphy.com/v1/gifs/search?api_key=9O1lQTPPrrsDfbowNS3VKp2CrO6q1bL8&q=DragonBall&limit=20'

    test('should match with snapshot', () => { 
        const { container } = render( <GifGridItems title={ title } url={ url } /> )
        expect( container ).toMatchSnapshot();
    })

    test('should display img with indicated alt and url ', () => { 
        render( <GifGridItems title={ title } url={ url } /> )
        // screen.debug(); 
        // expect( screen.getByRole('img').src ).toBe( url )
        const {src, alt} = screen.getByRole('img')
        expect( src ).toBe( url )
        expect( alt ).toBe( title )
    })

    test('should display title as a text <p></p> and not as a attribute', () => { 
        render( <GifGridItems title={ title } url={ url } /> )
        expect( screen.getByText(title) ).toBeTruthy();
    })
})