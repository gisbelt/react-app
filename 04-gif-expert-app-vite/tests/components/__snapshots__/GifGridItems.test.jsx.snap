// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component <GifGridItems /> should match with snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="DragonBall"
      src="https://api.giphy.com/v1/gifs/search?api_key=9O1lQTPPrrsDfbowNS3VKp2CrO6q1bL8&q=DragonBall&limit=20"
    />
    <p>
      DragonBall
    </p>
  </div>
</div>
`;
